central/auth/service/service_impl_test.go

```diff
@@ -126,3 +182,464 @@ func (s *authServiceAccessControlTestSuite) TestAuthServiceResponse() {
 		})
 	}
 }
+
+func (s *authServiceAccessControlTestSuite) TestValidateAuthMachineToMachineConfig() {
+	testCases := []struct {
+		config      *v1.AuthMachineToMachineConfig
+		skipIDCheck bool
+		err         error
+	}{
+		{
+			config: nil,
+			err:    errox.InvalidArgs,
+		},
+		{
+			config: &v1.AuthMachineToMachineConfig{},
+			err:    errox.InvalidArgs,
+		},
+		{
+			config: &v1.AuthMachineToMachineConfig{
+				Id: "some-id",
+			},
+			err: errox.InvalidArgs,
+		},
+		{
+			config: &v1.AuthMachineToMachineConfig{
+				Id:                      "some-id",
+				TokenExpirationDuration: "1",
+			},
+			err: errInvalidToken,
+		},
+		{
+			config: &v1.AuthMachineToMachineConfig{
+				Id:                      "some-id",
+				TokenExpirationDuration: "1s",
+			},
+			err: errox.InvalidArgs,
+		},
+		{
+			config: &v1.AuthMachineToMachineConfig{
+				Id:                      "some-id",
+				TokenExpirationDuration: "24h1s",
+			},
+			err: errox.InvalidArgs,
+		},
+		{
+			config: &v1.AuthMachineToMachineConfig{
+				Id:                      "some-id",
+				TokenExpirationDuration: "5m",
+				Type:                    v1.AuthMachineToMachineConfig_GENERIC,
+			},
+			err: errox.InvalidArgs,
+		},
+		{
+			config: &v1.AuthMachineToMachineConfig{
+				Id:                      "some-id",
+				TokenExpirationDuration: "5m",
+				Type:                    v1.AuthMachineToMachineConfig_GENERIC,
+			},
+			err: errox.InvalidArgs,
+		},
+		{
+			config: &v1.AuthMachineToMachineConfig{
+				Id:                      "some-id",
+				TokenExpirationDuration: "5m",
+				Mappings: []*v1.AuthMachineToMachineConfig_Mapping{
+					{
+						Key:   "some-key",
+						Value: "some-value",
+						Role:  testRole1,
+					},
+					{
+						Key:   "some-key",
+						Value: "some-value",
+						Role:  testRole2,
+					},
+					{
+						Key:   "some-key",
+						Value: "some-value",
+						Role:  "This doesn't exist",
+					},
+				},
+			},
+			err: errox.InvalidArgs,
+		},
+		{
+			config: &v1.AuthMachineToMachineConfig{
+				Id:                      "some-id",
+				TokenExpirationDuration: "5m",
+				Issuer:                  "something-invalid/%+o",
+			},
+			err: errInvalidIssuer,
```

Comment:
Just for me to confirm: is issuer always guaranteed to be URL? I was assuming so, but the rfc doesn't seem to prove that <https://datatracker.ietf.org/doc/html/rfc7519#section-4.1.1>.

Also, interestingly, it says

```
    StringOrURI
 A JSON string value, with the additional requirement that while
 arbitrary string values MAY be used, any value containing a ":"
 character MUST be a URI [RFC3986].  StringOrURI values are
 compared as case-sensitive strings with no transformations or
 canonicalizations applied.
```

Reply:
[Yes according to the OIDC spec.](https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata)